packages/backend/server/src/plugins/copilot/providers/types.ts

```diff
@@ -107,81 +102,44 @@ export const CopilotImageOptionsSchema = CopilotProviderOptionsSchema.merge(
 
 export type CopilotImageOptions = z.infer<typeof CopilotImageOptionsSchema>;
 
-export interface CopilotTextToTextProvider extends CopilotProvider {
-  generateText(
-    messages: PromptMessage[],
-    model: string,
-    options?: CopilotChatOptions
-  ): Promise<string>;
-  generateTextStream(
-    messages: PromptMessage[],
-    model: string,
-    options?: CopilotChatOptions
-  ): AsyncIterable<string>;
-}
+export const CopilotEmbeddingOptionsSchema =
+  CopilotProviderOptionsSchema.extend({
+    dimensions: z.number(),
+  }).optional();
 
-export interface CopilotTextToEmbeddingProvider extends CopilotProvider {
-  generateEmbedding(
-    messages: string[] | string,
-    model: string,
-    options?: CopilotEmbeddingOptions
-  ): Promise<number[][]>;
+export type CopilotEmbeddingOptions = z.infer<
+  typeof CopilotEmbeddingOptionsSchema
+>;
+
+export enum ModelInputType {
+  Text = 'text',
+  Image = 'image',
+  Audio = 'audio',
 }
 
-export interface CopilotTextToImageProvider extends CopilotProvider {
-  generateImages(
-    messages: PromptMessage[],
-    model: string,
-    options?: CopilotImageOptions
-  ): Promise<Array<string>>;
-  generateImagesStream(
-    messages: PromptMessage[],
-    model: string,
-    options?: CopilotImageOptions
-  ): AsyncIterable<string>;
+export enum ModelOutputType {
+  Text = 'text',
+  Embedding = 'embedding',
+  Image = 'image',
+  Structured = 'structured',
 }
 
-export interface CopilotImageToTextProvider extends CopilotProvider {
-  generateText(
-    messages: PromptMessage[],
-    model: string,
-    options: CopilotChatOptions
-  ): Promise<string>;
-  generateTextStream(
-    messages: PromptMessage[],
-    model: string,
-    options: CopilotChatOptions
-  ): AsyncIterable<string>;
+export interface ModelCapability {
+  input: ModelInputType[];
+  output: ModelOutputType[];
+  defaultForOutputType?: boolean;
 }
 
-export interface CopilotImageToImageProvider extends CopilotProvider {
-  generateImages(
-    messages: PromptMessage[],
-    model: string,
-    options?: CopilotImageOptions
-  ): Promise<Array<string>>;
-  generateImagesStream(
-    messages: PromptMessage[],
-    model: string,
-    options?: CopilotImageOptions
-  ): AsyncIterable<string>;
+export interface CopilotProviderModel {
+  id: string;
+  capabilities: ModelCapability[];
 }
 
-export type CapabilityToCopilotProvider = {
-  [CopilotCapability.TextToText]: CopilotTextToTextProvider;
-  [CopilotCapability.TextToEmbedding]: CopilotTextToEmbeddingProvider;
-  [CopilotCapability.TextToImage]: CopilotTextToImageProvider;
-  [CopilotCapability.ImageToText]: CopilotImageToTextProvider;
-  [CopilotCapability.ImageToImage]: CopilotImageToImageProvider;
+export type ModelConditions = {
+  inputType?: ModelInputType;
```

Comment:
inputType?: ModelInputType[];

Because in multimodal scenarios, users can enter text and pictures at the same time.

Reply:
done